### PR TITLE
Task/14812 Preparation Dev menu support to change the date

### PR DESCRIPTION
**Pull request by flxschmidt**
## Description
New Developer Menu Feature: `CWA Hibernation` Options. Set the comparing date to trigger hibernation status as you like.

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-14812

## Screenshots
Screenshot 2023-02-17 at 16 06 07

### CODE REVIEW

File: src/xcode/ENA/ENA.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -14,6 +14,7 @@
 		0104CBCF273EAC5B009A543D /* TicketValidationResultToken.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0104CBCE273EAC5B009A543D /* TicketValidationResultToken.swift */; };
 		0104CBD1273EAC7A009A543D /* TicketValidationResultToken+Fake.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0104CBD0273EAC7A009A543D /* TicketValidationResultToken+Fake.swift */; };
 		0104CBD3273EACA1009A543D /* MockTicketValidation.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0104CBD2273EACA1009A543D /* MockTicketValidation.swift */; };
+        0104CBD5273EAD72009A543D /* TicketValidation.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0104CBD4273EAD72009A543D /* TicketValidation.swift */; };
```
Comment (30mar) [inline src/xcode/ENA/ENA.xcodeproj/project.pbxproj:17]: is this duplicated or moved or what?  the PR has nothing to do with TicketValidation

File: src/xcode/ENA/ENA/Source/Workers/Store/SecureStore.swift [hunk 2]
```diff
@@ -377,6 +377,11 @@ final class SecureStore: SecureKeyValueStoring, Store {
 		get { kvStore["recentTraceLocationCheckedInto"] as DMRecentTraceLocationCheckedInto? ?? nil }
 		set { kvStore["recentTraceLocationCheckedInto"] = newValue }
 	}
+	
+	var hibernationComparingDate: Date {
```
Comment (30mar) [inline src/xcode/ENA/ENA/Source/Workers/Store/SecureStore.swift]: I think we should the if !RELEASE here and not in the initializer for the CWAHibernationProvider, we already did this several times, for example: elsLoggingActiveAtStartup in SecureStore

**Review by 30mar — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Workers/Store/SecureStore.swift (hunk 2 above)
Comment (flxschmidt) [reply src/xcode/ENA/ENA/Source/Workers/Store/SecureStore.swift]: Please see line `354.` `hibernationComparingDate` is in the `!RELEASE` block. ;)

File: src/xcode/ENA/ENA.xcodeproj/project.pbxproj (hunk 1 above)
Comment (flxschmidt) [reply src/xcode/ENA/ENA.xcodeproj/project.pbxproj:17]: Good point, don't know what Xcode has done here automatically. There is no corresponding change to TicketValidation as you wrote. 🤔

**Review by 30mar — APPROVED** [review]

**Review by puneetmahali1 — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/AppDelegate & Globals/CWAHibernationProvider.swift [hunk 3]
```diff
@@ -11,7 +11,14 @@ class CWAHibernationProvider: RequiresAppDependencies {
 	/// Use shared instance instead
 	private init() {}
 	
-	// MARK: - Internal
+	#if !RELEASE
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/AppDelegate & Globals/CWAHibernationProvider.swift]: Please use #if DEBUG

!RELEASE != DEBUG and hence the pipeline fails.

File: src/xcode/ENA/ENA/Source/AppDelegate & Globals/CWAHibernationProvider.swift [hunk 4]
```diff
@@ -20,7 +27,7 @@ class CWAHibernationProvider: RequiresAppDependencies {
 		#if RELEASE
 		return Date() >= hibernationStartDate
 		#else
-		return true // to.do Dev Menu setting, see https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-14812
+		return secureStore.hibernationComparingDate >= hibernationStartDate
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/AppDelegate & Globals/CWAHibernationProvider.swift]: My Suggestion hibernationComparisonDate or HibernationDate, please change it other places as well.

```suggestion
		return secureStore.hibernationComparisonDate >= hibernationStartDate
```

File: src/xcode/ENA/ENA/Source/AppDelegate & Globals/CWAHibernationProvider.swift [hunk 5]
```diff
@@ -41,4 +48,17 @@ class CWAHibernationProvider: RequiresAppDependencies {
...
+	
+	private var secureStore: Store {
+		#if RELEASE
+		return store
+		#else
+		return customStore ?? store
+		#endif
+	}
+	
+	#if !RELEASE
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/AppDelegate & Globals/CWAHibernationProvider.swift]: Please use #if DEBUG

**Review by naveeddotio — CHANGES_REQUESTED** [review]
Please see my comments

File: src/xcode/ENA/ENA/Source/Developer Menu/DMViewController.swift [hunk 6]
```diff
@@ -183,6 +179,8 @@ final class DMViewController: UITableViewController, RequiresAppDependencies {
 			vc = DMCrashAppViewController()
 		case .srs:
 			vc = DMSRSOptionsViewController(store: store)
+		case .hibernation2023:
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Developer Menu/DMViewController.swift]:
```suggestion
		case .hibernation:
```

File: src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,91 @@
+//
+// 🦠 Corona-Warn-App
+//
+
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift:4]: Missing #if !RELEASE ?

File: src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+	func numberOfRows(in section: Int) -> Int { 1 }
+	
+	func titleForFooter(in section: Int) -> String {
+		guard let section = Sections(rawValue: section) else {
+			fatalError("Invalid tableView section")
+		}
+		
+		switch section {
+		case .hibernationComparingDate:
+			var title = "App will be shutdown after set a new date value in the date picker.\n\nCurrently the hibernation threshold compares against the set date \(dateFormatter.string(from: store.hibernationComparingDate))"
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift]:
```suggestion
			var title = "App will shutdown after selecting a new date in the date picker.\n\nCurrently the hibernation threshold compares against the date: \(dateFormatter.string(from: store.hibernationComparingDate))"
```

File: src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+			
+			return title
+		case .reset:
+			return "App will be shutdown after reset to today date."
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift]:
```suggestion
			return "App will shutdown after reseting to today's date."
```

File: src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+	
+	func cellViewModel(for indexPath: IndexPath) -> Any {
+		guard let section = Sections(rawValue: indexPath.section) else {
+			fatalError("Invalid tableView section")
+		}
+		
+		switch section {
+		case .hibernationComparingDate:
+			return DMDatePickerCellViewModel(
+				title: "Hibernation Comparing Date",
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift]:
```suggestion
				title: "Hibernation Comparison Date",
```

File: src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+				accessibilityIdentifier: AccessibilityIdentifiers.DeveloperMenu.Hibernation.datePicker,
+				datePickerMode: .date,
+				date: store.hibernationComparingDate
+			)
+		case .reset:
+			return DMButtonCellViewModel(
+				text: "Reset Comparing Date",
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift]:
```suggestion
				text: "Reset Comparison Date",
```

File: src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+				textColor: .white,
+				backgroundColor: .enaColor(for: .buttonDestructive)
+			) { [weak self] in
+				self?.store(hibernationComparingDate: Date())
+			}
+		}
+	}
+	
+	func store(hibernationComparingDate: Date) {
+		Log.debug("[Debug-Menu] Set hibernation comparing date to \(dateFormatter.string(from: hibernationComparingDate)).")
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift]:
```suggestion
		Log.debug("[Debug-Menu] Set hibernation comparison date to: \(dateFormatter.string(from: hibernationComparingDate)).")
```

File: src/xcode/ENA/ENA/Source/Developer Menu/Helper/DMMenuItem.swift [hunk 13]
```diff
@@ -101,6 +102,7 @@ extension DMMenuItem {
 		case .boosterRules: return "Download Booster rules"
 		case .crashApp: return "Crash App"
 		case .srs: return "SRS Options"
+		case .hibernation2023: return "CWA Hibernation 2023"
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Developer Menu/Helper/DMMenuItem.swift]:
```suggestion
		case .hibernation: return "CWA Hibernation"
```

File: src/xcode/ENA/ENA/Source/AppDelegate & Globals/CWAHibernationProvider.swift (hunk 3 above)
Comment (flxschmidt) [reply src/xcode/ENA/ENA/Source/AppDelegate & Globals/CWAHibernationProvider.swift]: Good catch, changed.

File: src/xcode/ENA/ENA/Source/AppDelegate & Globals/CWAHibernationProvider.swift (hunk 4 above)
Comment (flxschmidt) [reply src/xcode/ENA/ENA/Source/AppDelegate & Globals/CWAHibernationProvider.swift]: adapted.

File: src/xcode/ENA/ENA/Source/AppDelegate & Globals/CWAHibernationProvider.swift (hunk 5 above)
Comment (flxschmidt) [reply src/xcode/ENA/ENA/Source/AppDelegate & Globals/CWAHibernationProvider.swift]: done.

File: src/xcode/ENA/ENA/Source/Developer Menu/DMViewController.swift (hunk 6 above)
Comment (flxschmidt) [reply src/xcode/ENA/ENA/Source/Developer Menu/DMViewController.swift]: changed.

File: src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift (hunk 7 above)
Comment (flxschmidt) [reply src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift:4]: added

File: src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift (hunk 8 above)
Comment (flxschmidt) [reply src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift]: changed.

File: src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift (hunk 9 above)
Comment (flxschmidt) [reply src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift]: changed.

File: src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift (hunk 10 above)
Comment (flxschmidt) [reply src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift]: changed.

File: src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift (hunk 11 above)
Comment (flxschmidt) [reply src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift]: changed.

File: src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift (hunk 12 above)
Comment (flxschmidt) [reply src/xcode/ENA/ENA/Source/Developer Menu/Features/DMHibernation/DMHibernationOptionsViewModel.swift]: changed.

File: src/xcode/ENA/ENA/Source/Developer Menu/Helper/DMMenuItem.swift (hunk 13 above)
Comment (flxschmidt) [reply src/xcode/ENA/ENA/Source/Developer Menu/Helper/DMMenuItem.swift]: changed.

**Review by naveeddotio — APPROVED** [review]

**Pull request merged**